Expand the doc comment for Context

natebosch opened the pull request:
This is the main point of interaction for authors of extensions, so docs
cover a wide range of information, including how the descriptions will
be used in failure output. Include some descriptions about how behavior
can vary across different contexts, and some high level descriptions
of the behavior for the contexts related to specific utilities.

lrhn reviewed (approved):
Lots of useful information, but I'd prefer some more conceptual stuff up-front, to set the stage for the technical details below. (Introduce concepts before referring to them, so that I know the what and why about them.)

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
+/// strictly after the expectation callback is called.
+///
+/// Some expectations check something about the subject, and do not allow for
+/// further checking. These expectations use [expect], [expectAsync], or
```
> It suggests that expectations either use `nest*` or `expect*`, but not both.
> I'd expect some expectation extension methods to do both, like first check that `length` is 1, then next `first`, to implement a checking version of `single`.

I don't expect any expectations extensions to do both. If `single` were implemented without a `single` member on the actual, I would expect it to look like

```dart
  Subject<T> get single => context.nest(() => ['has single element'], (actual) {
        if (actual.length != 1) {
          return Extracted.rejection(which: ['does not have length 1']);
        }
        return Extracted.value(actual.first);
      });
```

The reason to use a single call on `context` is that a single call corresponds to single description. I will maybe talk about how a call to `expect` or `nest` can be combined with other expectations, or how multiple expectation extensions could be called in a single "expectation combining extension" the difference is that when combining expectations, the failure description will have all the descriptions, instead of a single description which covers the whole expectation.

```dart
  Subject<T> get single {
    this.length.equals(1);
    return this.first;
  }
```

lrhn replied:
Makes sense.

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
 ///
-/// This is the surface of interaction for expectation extension method
-/// implementations.
+/// Expectation extension methods use [ContextExtension] to get access to the
```
> Is there some overview documentation (probably elsewhere) which describes the general concepts and their relations, which can be linked to before going into details?

Not currently. Where do you think the best place to put that would be? The README?

lrhn replied:
That, and/or the library documentation for the public libraries.

And for each important concept, describe itself thoroughly in its documentation, and then mention and link to the other relevant concepts, enough to show their relation.

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
 ///
-/// This is the surface of interaction for expectation extension method
-/// implementations.
+/// Expectation extension methods use [ContextExtension] to get access to the
```
This is super helpful, and I agree it would be more clear to reorder the information.

natebosch replied:
Ok, I didn't use the exact order you suggested, but I used this structure for inspiration and I do think the doc is easier to read through now.

WDYT?

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
+/// For example, if a failure happens on the root subject, the "actual" is taken
+/// directly from the rejection.
+///
+///   Expected: a Future<int> that:
```
Done.

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
+/// used and the callbacks will not be called. Where possible, string formatting
+/// for the descriptions should be performed in the callback, not ahead of time.
+///
+/// When an expectation is used on a [ConditionSubject] and passed to
```
I rephrased this so it's hopefully more clear.

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
-/// implementations.
+/// Expectation extension methods use [ContextExtension] to get access to the
+/// subject's [Context]. Expectations pass callbacks which describe what is
+/// checked, and perform the expectation, respectively.
```
I think I fixed it so that I consistently use "expectation extension method" (or in a few places "expectation extension"), and the term "expectation" should only be used in the more general "a thing that can be expected of a subject" sense.

In the user facing docs describing writing tests, I suspect it will feel natural to describe calling an expectation extension method on a subject as "checking an expectation" on the subject, or for a `ConditionSubject` "recording and expectation".

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
+/// description callbacks are used.
+///
+/// When both callbacks are used, the description callback will always be called
+/// strictly after the expectation callback is called.
```
> Do you have, or expect, code which depends on that order?

Yes. Sometimes where I can't do `async` work for the description callback, I might have a fallback description by default, but gather some async information during the predicate and format it for a richer description when possible.

For example the `inOrder` extension method cannot synchronously describe the `Condition` instances it will check because they will use async expectations. It has a fallback for `describe` calling out the number of conditions that need to be satisfied, but only shows their details when the actual has run.

https://github.com/dart-lang/test/blob/911659bbc3eea9c133263b2cfcaacdc376767faf/pkgs/checks/lib/src/extensions/async.dart#L221-L224

> Such code would likely maintain internal state which makes it impossible to replay the same check/description on a separate value, which would make me very disappointed.

That is the case. We should investigate whether there is a different way to provide a good UX for this.
Maybe we can devise an explicit way for the expectation callback to communicate to the description callback that would be compatible with reuse. Or maybe we should sacrifice the UX for those use cases.

natebosch commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
+///   Expected a Future<int> that:
+///     completes to a value
+///
+/// Expectations which call [expectAsync] or [nestAsync] may not be used in the
```
Done.

jakemac53 commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,40 +187,204 @@ extension ContextExtension<T> on Subject<T> {
-/// The `nest` and `nestAsync` can test the value, and also extract some other
-/// property from it for further checking.
+/// The user focused APIs called within tests are expectation extension methods
+/// written in extension `on Subject`, typically specialized to a specific
```
```suggestion
/// written as extensions `on Subject`, typically specialized to a specific
```

natebosch replied:
Changed to `in an extension`

jakemac53 commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,40 +187,204 @@ extension ContextExtension<T> on Subject<T> {
+/// Expectation extension methods will make a call to one of the APIs on the
+/// subject's [Context], and can perform one of two types of operations:
+///
+/// -   Expect something of the current value.
+/// -   Expect that a new subject can be extracted from the current value.
```
Would it be worth showing an example usage of each of these here, just to drive the point home a bit?
```suggestion
/// -   Expect something of the current value (`checkThat(1).equals(1)`).
/// -   Expect that a new subject can be extracted from the current value (`checkThat([]).length`).
```
Or you could have direct links to some real ones here?

natebosch replied:
> Or you could have direct links to some real ones here?

It would introduce circular imports. I guess that wouldn't be a big deal within `src`...

natebosch replied:
I went ahead and added the links, but they don't work currently.

https://github.com/dart-lang/dartdoc/issues/3325

jakemac53 commented on pkgs/checks/lib/src/checks.dart:
```diff
@@ -187,40 +187,204 @@ extension ContextExtension<T> on Subject<T> {
+///
+///
+/// Whichever type of operation, an expectation extension method provides two
+/// callbacks.
```
I think some sort of link to the api  here would be good, to show how these are provided, it's a bit confusing without that imo. I end up wanting to interrupt by reading to find that.

natebosch replied:
I expanded this to show one callback signature, and describe that the other signature varies. Does that help?

jakemac53 replied:
What I am confused by here isn't what the callbacks are supposed to do, its how/where they are provided.

I think this stems from the mentions of `expect` etc being below this - these are describing how those work right? Feels like its putting the cart before the horse a bit, I don't have the context yet for what is being described.

natebosch replied:
How about mentioning each method in the bullet list above of the two types of operations?

jakemac53 reviewed (commented):
In general I think my main comment would be that there is no example code in these docs, or links to examples.

I think a lot of people learn best by example, and just linking to existing (built-in) extensions from several of the sections, or providing really basic examples inline, would help a lot.

natebosch commented:
> or providing really basic examples inline

I was a little worried about the docs getting too long. I also don't know how much we want to repeat between the README and the API docs - maybe we should be OK with copy pasting a bunch between them?

Would it work to add an example usage similar to what is in the README on each of the individual APIs, and leave the class level doc as is?

jakemac53 commented:
> I was a little worried about the docs getting too long. I also don't know how much we want to repeat between the README and the API docs - maybe we should be OK with copy pasting a bunch between them?
> 
> Would it work to add an example usage similar to what is in the README on each of the individual APIs, and leave the class level doc as is?

I think copy/pasting between them is fine. I do think the class level doc here should have examples, or link to examples.

natebosch commented:
> I do think the class level doc here should have examples, or link to examples.

I added a couple sync examples, one of expect and one of nest. It is towards the end of the doc. I added examples of all operations to the respective docs.

I also moved the detail about when callbacks when/won't be called towards the end of the doc. It's probably a detail most folks won't care about - it motivates the design but isn't important to know to use the design. It also now follows another mention of `softCheck` which I think makes it less confusing.
